typescript/infra/src/funding/adapters/SealevelFundingAdapter.ts

```diff
@@ -0,0 +1,304 @@
+    const delta = desiredBalanceLamports - currentBalanceLamports;
+    const minDelta =
+      (desiredBalanceLamports * BigInt(fundingThresholdFactor * 100)) / 100n;
+    // Determine if funding is needed
+    const fundingAmount = delta > minDelta ? delta : 0n;
+
+    this.logger.debug(
+      {
+        chain: this.chainName,
+        currentBalance: currentBalanceLamports.toString(),
+        desiredBalance: desiredBalanceLamports.toString(),
+        delta: delta.toString(),
+        minDelta: minDelta.toString(),
+        fundingThresholdFactor,
+        role,
+      },
+      'Funding amount',
+    );
+    return fundingAmount;
+  }
+
+  async fundKey(
+    key: BaseAgentKey,
+    desiredBalance: number,
+    fundingThresholdFactor: number,
+  ): Promise<void> {
+    const fundingAmountLamports = await this.getFundingAmount(
+      key.address,
+      desiredBalance,
+      fundingThresholdFactor,
+      key.role as FundableRole,
+    );
+
+    if (fundingAmountLamports === 0n) {
+      this.logger.info(
+        {
+          key: key.address,
+          chain: this.chainName,
+          role: key.role,
+        },
+        'Skipping funding for key',
+      );
+      return;
+    }
+
+    const fundingAmountSol = fundingAmountLamports / BigInt(LAMPORTS_PER_SOL);
+    const funderAddress = this.getSigner().publicKey.toBase58();
+
+    const funderBalanceLamports = await this.getBalance(funderAddress);
+    const funderBalanceSol = funderBalanceLamports / BigInt(LAMPORTS_PER_SOL);
+
+    const fromPubkey = new PublicKey(funderAddress);
+    const toPubkey = new PublicKey(key.address);
+
+    this.logger.info(
+      {
+        chain: this.chainName,
+        amount: fundingAmountSol.toString(),
+        key: key.address,
+        funder: {
+          address: funderAddress,
+          balance: funderBalanceSol.toString(),
+        },
+        context: this.context,
+      },
+      'Funding key',
+    );
+
+    try {
+      // Get latest blockhash for transaction
+      const { blockhash, lastValidBlockHeight } =
+        await this.getProvider().getLatestBlockhash('finalized');
+
+      // Create transaction
+      const transaction = new Transaction({
+        feePayer: fromPubkey,
+        blockhash,
+        lastValidBlockHeight,
+      });
+
+      // Add transfer instruction
+      transaction.add(
+        SystemProgram.transfer({
+          fromPubkey,
+          toPubkey,
+          lamports: fundingAmountLamports,
+        }),
+      );
+
+      // Add priority fee if configured for this chain
+      const priorityFee = SEALEVEL_PRIORITY_FEES[this.chainName];
+      if (priorityFee) {
+        transaction.add(
+          ComputeBudgetProgram.setComputeUnitPrice({
+            microLamports: priorityFee,
+          }),
+        );
+      }
+
+      // Send and confirm transaction
+      const signature = await sendAndConfirmTransaction(
+        this.getProvider(),
+        transaction,
+        [this.getSigner()],
+      );
```

Comment:
Can we have some retry logic here, solana is quite messy when trying to land txs, like if the error is `transaction blockchash expired` we retry otherwise we just throw